drivers/hmis/app/models/hmis/ce/filtered_candidates_query.rb

```diff
@@ -0,0 +1,111 @@
+###
+# Copyright 2016 - 2025 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# frozen_string_literal: true
+
+module Hmis::Ce
+  # Query for Coordinated Entry opportunity candidates with an option to filter
+  # clients who have recently declined a referral.
+  #
+  # This filter identifies clients who have been declined from any opportunity within
+  # the same unit group as the target opportunity.
+  #
+  # An exception is made for clients who have been re-assessed since their most
+  # recent decline. Since a re-assessment may change their eligibility, they are
+  # included in the candidate list regardless of the recent decline.
+  #
+  class FilteredCandidatesQuery
+    def arel = Hmis::ArelHelper.instance
+
+    def initialize(opportunity:, exclude_recently_declined: false)
+      @opportunity = opportunity
+      @unit_group_id = opportunity.unit_group.id
+      @exclude_recently_declined = exclude_recently_declined
+      @base_candidate_scope = Hmis::Ce::Match::Candidate.for_opportunity(opportunity).prioritized
+    end
+
+    def resolve
+      return @base_candidate_scope unless @exclude_recently_declined
+
+      # { source_client_id => most_recent_decline_timestamp }
+      most_recent_declines = fetch_most_recent_declines
+      return @base_candidate_scope if most_recent_declines.blank?
+
+      # { source_client_id => dest_client_id }
+      source_to_dest_client_map = map_source_to_dest_clients(most_recent_declines.keys)
+      return @base_candidate_scope if source_to_dest_client_map.blank? # unexpected
```

Comment:
I think this would occur if no clients have been declined from the unit group, because then `most_recent_declines` would be empty -- so this doesn't seem unexpected, if I'm understanding correctly